# Problem Sets

Here you will find the problem sets that accompany each module of the course. The first three problem sets are designed to take two weeks to complete and the last problem set is designed to take three weeks. Problem sets are designed to be completed individually by students and are a mix of analyzing public datasets, simulations, and derivations.



## Overview

### `ps1`

For this problem set studnets will look at patterns of genetic diversity at the *LCT* and *MCM6* genes in the phase three release of the 1000 Genomes Project. Studies have shown that these loci exhibit signals consistent with positive selection in European populations and is associated with lactase persistence—a trait unique among mammals thought to be a result of cattle domestication and the incorporation of milk into the adult diets of several human populations (Enattah et al. 2002; Marnetto and Huerta-Sánchez 2017; Smith et al. 2018).

### `ps2`

For this problem set students will validate classical derivations of tree attributes from lecture using coalescent simulations in `msprime` and practice manipulating tree-sequences using `tskit`.

### `ps3`

For this problem students will first derive the expected branch length for the *BBAA* site pattern from Peede et al. 202X (_in-prep_) and then validate the expected branch length derivations for the _ABBA_, _BABA_, and _BBAA_ site patterns using coalescent simulations in `msprime`.

### `ps4`

*Heliconius* butterflies are isolated by geography and display striking differences in adaptive wing patterns, which is facilitated by Müllerian mimicry between distantly related species and coupled with divergence between closely related species. *Heliconius* butterflies frequently hybridize when their respective species ranges overlap and it has been hypothesized that introgression might play an important role in speciation and adaptive radiation. For the final problem set students will first look at global patterns of population structure and differentiation between *Heliconius* butterflies subspecies and then assess claims about natural selection and introgression at the *optix* gene—a gene that controls the extreme red wing pattern variation across multiple species of *Heliconius* and has been implicated as one of the central genes involved in wing pattern mimicry (Jiggins et al. 2017). To accomplish this students will analyze the chromosome 18 data from Martin et al. 2020 which has been already pre-filtered for students and contains some missing data—i.e. students will need to properly account for missing data throughout all analyses in this problem set.